### PR TITLE
feat(tools): categorize ast_edit validation errors as pre-existing vs newly-introduced (Fixes #2124)

### LINKED ISSUE #2124: AST "preexisting parse errors"
### What happened?

The llm claims the ast tool choked on some C++

[glm]
 The AST validation failed (it's a heuristic on a large file with some pre-existing weirdness — the "Syntax error at line 173" is likely a false positive from the tool's parsing of this 5400-line file, not my edit). The tool suggests calling with force:true. Since my edit is a trivial, safe insertion (a few printf lines at a function boundary), let me apply it with force.

Actually, let me reconsider — the AST error "Syntax error at line 173" is about the EXISTING file (line 173 is far from my edit at 4977). This is the tool's parser struggling with the large/unusual file. My edit is straightforward. Let me force it.

Wait — actually, let me use the simpler `replace` tool instead, which does exact text matching without AST validation. That's safer for this. Let me use `replace`.

Actually, the ast_edit already showed a preview but didn't apply. Let me just use the `replace` tool for a clean exact-text replacement.

 The AST validator choked on the large file (a pre-existing parse quirk at line173, unrelated to my edit). My change is a trivial insertion at a function boundary, so let me apply it with the exact-text tool instead.

### What did you expect to happen?

ast should attempt to limit the parse to relevant code and thereafter -- I wonder if this was what the llm said we should investigate it.

### Client information

About LLxprt Code

 CLI Version                                                   0.10.0-nightly.260622.487a5b119
 Git Commit                                                    487a5b119

 Sandbox                                                       no sandbox
 OS                                                            darwin

### Login information

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by acoliver**
## Summary

The `ast_edit` tool validated only post-edit content, so any pre-existing syntax error in the file looked like the edit's fault. This caused the LLM to either blindly `force: true` the edit or abandon `ast_edit` for the simpler `replace` tool, as described in issue #2124.

This PR adds **pre-edit baseline validation** and a **categorization layer** that compares the pre-edit and post-edit parse trees to tell the LLM whether a post-edit validation failure was already there.

### New: `validation-categorizer.ts`
A focused module that categorizes validation results:
- **Pre-existing**: post-edit error matches a pre-edit error at the same (or shifted) location
- **Newly-introduced**: pre-edit was clean but post-edit has errors
- **Mixed**: file had pre-existing errors but post-edit errors are at different locations

The categorizer uses the edit's **start line** and **net line delta** to deterministically match shifted error locations. Errors at or below the edit must have shifted by `lineDelta`; errors above did not shift. This avoids false positives where a newly-introduced error is mistaken for a pre-existing one.

### Modified: `edit-calculator.ts`
- `calculateEdit()` now runs `validateASTSyntax()` on the original content (before the replacement) and returns it as `preEditValidation` alongside the existing `astValidation`.

### Modified: `ast-edit-invocation.ts`
- **Preview output**: shows categorized labels like `FAILED (pre-existing error at line 173 — present before this edit)` or `FAILED (new error introduced by this edit at line 42)`
- **Apply output**: same categorization in the success message
- **Confirmation metadata**: includes `preEditValidation` alongside `astValidation`

## Example output

**Before (issue scenario):**
```
- AST validation: FAILED
- AST errors: Syntax error at line 173, column 5
```

**After:**
```
- AST validation: FAILED (pre-existing error at line 173 — present before this edit)
- Pre-existing syntax errors: Yes at line 173
- AST errors: Syntax error at line 173, column 5
```

## Testing

- 28 unit tests for the categorizer (tolerance boundaries, edit-start-line precision, line delta edge cases including trailing newlines)
- 10 integration tests exercising the full preview + apply flow for pre-existing, newly-introduced, and clean scenarios
- All 465 existing tools package tests still pass

## Notes

- The categorization is advisory (helps the LLM decide), not a hard gate — the raw error is always surfaced regardless
- The `SKIPPED` status is used when validation cannot run (e.g., unsupported language) rather than misleadingly reporting `PASSED`
- `computeLineDelta` trims a single trailing newline so that inserting a newline-only line does not inflate the shift estimate

Closes #2124

### CODE REVIEW

File: packages/tools/src/tools/ast-edit/validation-categorizer.ts [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+  // Both pre- and post-edit are invalid. Determine whether the post-edit errors
+  // correspond to the same locations as the pre-edit ones.
+  const preLines = extractErrorLineNumbers(preEdit.errors);
+  const postLines = extractErrorLineNumbers(postErrors);
+  const allMatch =
+    postLines.length > 0 &&
+    postLines.every((pl) =>
+      preLines.some((prel) => linesMatch(prel, pl, lineDelta, editStartLine)),
+    );
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/validation-categorizer.ts]: Fixed in commit aca6a834d. Rewrote the matching logic to iterate over ALL post-edit error strings (not just parseable ones). Any post-edit error that lacks a parseable line number is now conservatively treated as unmatched, correctly setting `newlyIntroduced: true`.

File: packages/tools/src/tools/ast-edit/__tests__/ast-edit-preview.test.ts [hunk 2]
```diff
@@ -165,3 +165,133 @@ describe('ASTEditTool preview phase validation (issue #1755)', () => {
     );
   });
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/__tests__/ast-edit-preview.test.ts:166]: Fixed in commit aca6a834d. Added a mixed-error integration test that creates a file with a pre-existing syntax error AND introduces a new error via the edit, verifying both are surfaced correctly.

File: packages/tools/src/tools/ast-edit/ast-edit-invocation.ts [hunk 3]
```diff
@@ -378,15 +425,39 @@ export class ASTEditToolInvocation
...
       };
 
+      const { lineDelta, editStartLine } =
+        this.computeValidationContext(editData);
+      const summary = summarizeAstValidation(
+        editData.preEditValidation,
+        editData.astValidation,
+        lineDelta,
+        editStartLine,
+      );
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/ast-edit-invocation.ts]: Fixed in commit aca6a834d. writeEditResult now destructures and uses all four values from computeValidationContext (astValidation, preEditValidation, lineDelta, editStartLine) instead of discarding them and re-reading raw editData fields. This preserves the ?? fallback logic.

File: packages/tools/src/tools/ast-edit/__tests__/validation-categorizer.test.ts [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+import { describe, it, expect } from 'vitest';
+import {
+  summarizeAstValidation,
+  computeLineDelta,
+  findEditStartLine,
+  extractErrorLineNumber,
+  formatValidationLineLabel,
+} from '../validation-categorizer.js';
+
+describe('summarizeAstValidation', () => {
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/__tests__/validation-categorizer.test.ts:16]: Fixed in commit aca6a834d. Added tests for: multiple pre-existing errors all matching, partial multi-error match (one matches, one does not), and unparseable error in post-edit array.

File: packages/tools/src/tools/ast-edit/__tests__/validation-categorizer.test.ts [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+
+  it('does not inflate delta for trailing newlines', () => {
+    // A trailing newline terminates a line; 'abc\n' is 1 line, not 2.
+    expect(computeLineDelta('abc\n', 'abc\n')).toBe(0);
+    expect(computeLineDelta('abc', 'abc\n')).toBe(0);
+    expect(computeLineDelta('abc\n', 'abc')).toBe(0);
+  });
+});
+
+describe('findEditStartLine', () => {
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/__tests__/validation-categorizer.test.ts:308]: Fixed in commit aca6a834d. Added test verifying findEditStartLine returns the first occurrence when oldString appears multiple times, documenting the current behavior.

File: packages/tools/src/tools/ast-edit/__tests__/validation-categorizer.test.ts [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+    expect(findEditStartLine('abc\ndef', 'xyz')).toBeNull();
+  });
+
+  it('returns null for null content or empty old_string', () => {
+    expect(findEditStartLine(null, 'abc')).toBeNull();
+    expect(findEditStartLine('abc', '')).toBeNull();
+  });
+});
+
+describe('extractErrorLineNumber', () => {
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/__tests__/validation-categorizer.test.ts:336]: Fixed in commit aca6a834d. Added tests for alternative formats: 'Error on line 42' and 'Parse error at line 99'. Note: the regex matches the 'line N' pattern which is what our validateASTSyntax produces; formats without 'line' (e.g., '173:5') intentionally do not match.

File: packages/tools/src/index.ts [hunk 7]
```diff
@@ -411,6 +411,15 @@ export { ASTConfig } from './tools/ast-edit/ast-config.js';
 export { ASTQueryExtractor } from './tools/ast-edit/ast-query-extractor.js';
 export { RepositoryContextProvider } from './tools/ast-edit/repository-context-provider.js';
 export { validateASTSyntax } from './tools/ast-edit/edit-calculator.js';
+export {
+  summarizeAstValidation,
+  computeLineDelta,
+  findEditStartLine,
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/index.ts:417]: Dismissed — false positive. Preview (force:false) and apply (force:true) are separate tool invocations with separate object instances and separate process lifecycles. There is no single execution path where computeValidationContext is called twice. The two-phase architecture means the apply invocation never has access to the preview invocation's results.

File: packages/tools/src/index.ts [hunk 8]
```diff
@@ -411,6 +411,15 @@ export { ASTConfig } from './tools/ast-edit/ast-config.js';
...
+  extractErrorLineNumber,
+  formatValidationLineLabel,
+  type AstValidationResult,
+  type AstValidationSummary,
+} from './tools/ast-edit/validation-categorizer.js';
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/index.ts:422]: Acknowledged as a known limitation. The ast_edit tool enforces single-replacement semantics (countOccurrences returns 0 or 1, and applyReplacement uses String.replace which replaces only the first match). So old_string must be unique for the edit to apply at all. findEditStartLine resolving to the first match is consistent with this behavior. A test documenting this was added in commit aca6a834d.

File: packages/tools/src/tools/ast-edit/__tests__/ast-edit-preview.test.ts [hunk 9]
```diff
@@ -165,3 +165,160 @@ describe('ASTEditTool preview phase validation (issue #1755)', () => {
...
+      })
+      .execute(new AbortController().signal);
+
+    expect(result.error).toBeUndefined();
+    const content = String(result.llmContent);
+    // The baseline was already broken.
+    expect(content).toContain('Pre-existing syntax errors: Yes');
+    // Post-edit still has errors (either pre-existing or newly introduced).
+    expect(content).toContain('FAILED');
+  });
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/__tests__/ast-edit-preview.test.ts:342]: Fixed in commit 8e57d95e6. Strengthened the mixed-error test to use a file with 5 lines of padding between the edit site (line 1) and the pre-existing error (line 7), ensuring the errors exceed the 3-line tolerance and trigger genuine mixed categorization. Added assertion verifying the output does NOT contain 'present before this edit' (the pure pre-existing label).

File: packages/tools/src/tools/ast-edit/ast-edit-invocation.ts [hunk 10]
```diff
@@ -232,20 +242,60 @@ export class ASTEditToolInvocation
...
+    editStartLine: number | null;
+  } {
+    const astValidation =
+      editData.astValidation ??
+      this.validateASTSyntax(this.params.file_path, editData.newContent);
+    const preEditValidation: AstValidationResult | undefined =
+      editData.preEditValidation ??
+      (editData.currentContent !== null
+        ? this.validateASTSyntax(this.params.file_path, editData.currentContent)
+        : undefined);
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/ast-edit-invocation.ts:258]: Dismissed — false positive. validateASTSyntax already has an internal try/catch that returns { valid: false, errors: [...] } on any exception. It cannot throw. See edit-calculator.ts lines 265-302.

[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/ast-edit-invocation.ts:258]: Dismissed — false positive. The CalculatedEdit interface types currentContent as string | null (never undefined). The !== null guard is correct for this type. TypeScript's type system guarantees this at compile time.

File: packages/tools/src/tools/ast-edit/validation-categorizer.ts [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+  editStartLine: number | null,
+): boolean {
+  if (editStartLine !== null) {
+    // Errors at or below the edit start shifted by lineDelta.
+    if (postLine >= editStartLine) {
+      return Math.abs(preLine - (postLine - lineDelta)) <= LINE_MATCH_TOLERANCE;
+    }
+    // Errors above the edit start did not shift.
+    return Math.abs(preLine - postLine) <= LINE_MATCH_TOLERANCE;
+  }
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/validation-categorizer.ts]: Fixed in commit 8e57d95e6 — this was a real bug. The condition checked postLine >= editStartLine, but whether the PRE-EDIT error was at/below the edit determines whether it shifted. Changed to preLine >= editStartLine. The OCR example (preLine=5, editStartLine=10, postLine=15, delta=10) now correctly returns false: preLine(5) < editStartLine(10) so no-shift case applies, and |5 - 15| = 10 > tolerance(3). Added a regression unit test for this exact scenario.

File: packages/tools/src/index.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/index.ts:422]: Updated — fixed in commit 1e1ed888a. findEditStartLine now returns null when oldString appears more than once, since indexOf would return the first match but the actual edit could target a different occurrence. Returning null causes the categorizer to fall back to the safer lineDelta-based matching. The test was updated to verify this null-return behavior.

File: packages/tools/src/tools/ast-edit/ast-edit-invocation.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/ast-edit-invocation.ts:258]: Updated — fixed in commit 1e1ed888a. Changed !== null to != null defensively. While the CalculatedEdit type currently declares currentContent as string | null, using != null future-proofs against type evolution and is a harmless one-character change.

File: packages/tools/src/tools/ast-edit/validation-categorizer.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/validation-categorizer.ts]: Re-reviewed against current dev/0.11.0 after PR #2485. The existing response/fix remains valid; the branch is now integrated and the AST categorizer has additional exact line/column and conservative-boundary tests in 13370cacaf.

File: packages/tools/src/tools/ast-edit/__tests__/ast-edit-preview.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/__tests__/ast-edit-preview.test.ts:166]: Re-reviewed against current dev/0.11.0 after PR #2485. The existing response/fix remains valid; the branch is now integrated and the AST categorizer has additional exact line/column and conservative-boundary tests in 13370cacaf.

File: packages/tools/src/tools/ast-edit/ast-edit-invocation.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/ast-edit-invocation.ts]: Re-reviewed against current dev/0.11.0 after PR #2485. The existing response/fix remains valid; the branch is now integrated and the AST categorizer has additional exact line/column and conservative-boundary tests in 13370cacaf.

File: packages/tools/src/tools/ast-edit/__tests__/validation-categorizer.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/__tests__/validation-categorizer.test.ts:16]: Re-reviewed against current dev/0.11.0 after PR #2485. The existing response/fix remains valid; the branch is now integrated and the AST categorizer has additional exact line/column and conservative-boundary tests in 13370cacaf.

File: packages/tools/src/tools/ast-edit/__tests__/validation-categorizer.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/__tests__/validation-categorizer.test.ts:308]: Re-reviewed against current dev/0.11.0 after PR #2485. The existing response/fix remains valid; the branch is now integrated and the AST categorizer has additional exact line/column and conservative-boundary tests in 13370cacaf.

File: packages/tools/src/tools/ast-edit/__tests__/validation-categorizer.test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/__tests__/validation-categorizer.test.ts:336]: Re-reviewed against current dev/0.11.0 after PR #2485. The existing response/fix remains valid; the branch is now integrated and the AST categorizer has additional exact line/column and conservative-boundary tests in 13370cacaf.

File: packages/tools/src/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/index.ts:417]: Re-reviewed against current dev/0.11.0 after PR #2485. The existing response/fix remains valid; the branch is now integrated and the AST categorizer has additional exact line/column and conservative-boundary tests in 13370cacaf.

File: packages/tools/src/index.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/index.ts:422]: Re-reviewed against current dev/0.11.0 after PR #2485. The existing response/fix remains valid; the branch is now integrated and the AST categorizer has additional exact line/column and conservative-boundary tests in 13370cacaf.

File: packages/tools/src/tools/ast-edit/__tests__/ast-edit-preview.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/__tests__/ast-edit-preview.test.ts:342]: Re-reviewed against current dev/0.11.0 after PR #2485. The existing response/fix remains valid; the branch is now integrated and the AST categorizer has additional exact line/column and conservative-boundary tests in 13370cacaf.

File: packages/tools/src/tools/ast-edit/ast-edit-invocation.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/ast-edit-invocation.ts:258]: Re-reviewed against current dev/0.11.0 after PR #2485. The existing response/fix remains valid; the branch is now integrated and the AST categorizer has additional exact line/column and conservative-boundary tests in 13370cacaf.

[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/ast-edit-invocation.ts:258]: Re-reviewed against current dev/0.11.0 after PR #2485. The existing response/fix remains valid; the branch is now integrated and the AST categorizer has additional exact line/column and conservative-boundary tests in 13370cacaf.

File: packages/tools/src/tools/ast-edit/validation-categorizer.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/validation-categorizer.ts]: Re-reviewed against current dev/0.11.0 after PR #2485. The existing response/fix remains valid; the branch is now integrated and the AST categorizer has additional exact line/column and conservative-boundary tests in 13370cacaf.

File: packages/tools/src/tools/ast-edit/__tests__/validation-categorizer.test.ts [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+      formatValidationLineLabel([
+        'Syntax error at line 10, column 1',
+        'Syntax error at line 20, column 1',
+      ]),
+    ).toBe(' at lines 10, 20');
+  });
+
+  it('returns empty string when no line numbers', () => {
+    expect(formatValidationLineLabel(['unknown error'])).toBe('');
+  });
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/__tests__/validation-categorizer.test.ts:371]: Addressed in 13370cacaf (with the core semantic changes split across aa193aa4c2 and 13370cacaf): added the empty-list case, standardized preview execution, removed newline shadowing/char-code usage, tightened exact mixed/pure output assertions, and made the empty-error fallback explicit and grammatical.

File: packages/tools/src/tools/ast-edit/__tests__/ast-edit-preview.test.ts [hunk 13]
```diff
@@ -165,3 +165,172 @@ describe('ASTEditTool preview phase validation (issue #1755)', () => {
...
+
+    const tool = new ASTEditTool(createFakeToolHost(tempDir));
+    const result = await tool
+      .build({
+        file_path: filePath,
+        old_string: 'const greeting = "hello";',
+        new_string: 'const greeting = @@@;',
+        force: false,
+      })
+      .execute(new AbortController().signal);
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/__tests__/ast-edit-preview.test.ts]: Addressed in 13370cacaf (with the core semantic changes split across aa193aa4c2 and 13370cacaf): added the empty-list case, standardized preview execution, removed newline shadowing/char-code usage, tightened exact mixed/pure output assertions, and made the empty-error fallback explicit and grammatical.

File: packages/tools/src/tools/ast-edit/__tests__/ast-edit-preview.test.ts [hunk 14]
```diff
@@ -165,3 +165,172 @@ describe('ASTEditTool preview phase validation (issue #1755)', () => {
     );
   });
 });
+
+const NL = String.fromCharCode(10);
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/__tests__/ast-edit-preview.test.ts]: Addressed in 13370cacaf (with the core semantic changes split across aa193aa4c2 and 13370cacaf): added the empty-list case, standardized preview execution, removed newline shadowing/char-code usage, tightened exact mixed/pure output assertions, and made the empty-error fallback explicit and grammatical.

File: packages/tools/src/tools/ast-edit/__tests__/ast-edit-preview.test.ts [hunk 15]
```diff
@@ -165,3 +165,172 @@ describe('ASTEditTool preview phase validation (issue #1755)', () => {
...
+    expect(result.error).toBeUndefined();
+    const content = String(result.llmContent);
+    // The baseline was already broken.
+    expect(content).toContain('Pre-existing syntax errors: Yes');
+    // Post-edit still has errors.
+    expect(content).toContain('FAILED');
+    // Must NOT be classified as purely pre-existing — the edit introduced a
+    // new error at a different location, so the output should not claim it
+    // was "present before this edit".
+    expect(content).not.toContain('present before this edit');
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/__tests__/ast-edit-preview.test.ts:341]: Addressed in 13370cacaf (with the core semantic changes split across aa193aa4c2 and 13370cacaf): added the empty-list case, standardized preview execution, removed newline shadowing/char-code usage, tightened exact mixed/pure output assertions, and made the empty-error fallback explicit and grammatical.

File: packages/tools/src/tools/ast-edit/__tests__/ast-edit-preview.test.ts [hunk 16]
```diff
@@ -165,3 +165,172 @@ describe('ASTEditTool preview phase validation (issue #1755)', () => {
...
+    const result = await executePreview(tool, {
+      file_path: filePath,
+      old_string: 'return `hello ${name}`;',
+      new_string: 'return `world ${name}`;',
+      force: false,
+    });
+
+    expect(result.error).toBeUndefined();
+    const content = String(result.llmContent);
+    expect(content).toContain('pre-existing');
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/__tests__/ast-edit-preview.test.ts]: Addressed in 13370cacaf (with the core semantic changes split across aa193aa4c2 and 13370cacaf): added the empty-list case, standardized preview execution, removed newline shadowing/char-code usage, tightened exact mixed/pure output assertions, and made the empty-error fallback explicit and grammatical.

File: packages/tools/src/tools/ast-edit/__tests__/ast-edit-preview.test.ts [hunk 17]
```diff
@@ -165,3 +165,172 @@ describe('ASTEditTool preview phase validation (issue #1755)', () => {
...
+
+function makeBrokenFile(): string {
+  return [
+    'export function greet(name: string): string {',
+    '  return `hello ${name}`;',
+    '}',
+    'const broken = @@@;',
+    '',
+  ].join(NL);
+}
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/__tests__/ast-edit-preview.test.ts]: Addressed in 13370cacaf (with the core semantic changes split across aa193aa4c2 and 13370cacaf): added the empty-list case, standardized preview execution, removed newline shadowing/char-code usage, tightened exact mixed/pure output assertions, and made the empty-error fallback explicit and grammatical.

File: packages/tools/src/tools/ast-edit/validation-categorizer.ts [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+
+  // File was already broken, but the post-edit errors don't line up with the
+  // pre-existing ones. They may be cascading from the same root cause or newly
+  // introduced; surface both for the LLM to judge.
+  return {
+    status: 'FAILED',
+    preExisting: true,
+    newlyIntroduced: true,
+    label: `FAILED (file had pre-existing errors${formatLineLabel(preEdit.errors)}; verify post-edit error${formatLineLabel(postErrors)} is not newly introduced)`,
+  };
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/validation-categorizer.ts:243]: Addressed in 13370cacaf (with the core semantic changes split across aa193aa4c2 and 13370cacaf): added the empty-list case, standardized preview execution, removed newline shadowing/char-code usage, tightened exact mixed/pure output assertions, and made the empty-error fallback explicit and grammatical.

File: packages/tools/src/tools/ast-edit/ast-edit-invocation.ts [hunk 19]
```diff
@@ -232,20 +242,60 @@ export class ASTEditToolInvocation
...
+    editStartLine: number | null;
+  } {
+    const astValidation =
+      editData.astValidation ??
+      this.validateASTSyntax(this.params.file_path, editData.newContent);
+    const preEditValidation: AstValidationResult | undefined =
+      editData.preEditValidation ??
+      (editData.currentContent != null
+        ? this.validateASTSyntax(this.params.file_path, editData.currentContent)
+        : undefined);
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/ast-edit-invocation.ts:258]: Fixed in db96b7c161: restored strict null comparison, consistent with the CalculatedEdit string | null contract.

File: packages/tools/src/tools/ast-edit/ast-edit-invocation.ts [hunk 20]
```diff
@@ -378,15 +425,39 @@ export class ASTEditToolInvocation
...
         applied: true,
-        metadata: { astValidation: editData.astValidation },
+        metadata: {
+          astValidation: editData.astValidation,
+          preEditValidation: editData.preEditValidation,
+        },
       };
 
+      const { astValidation, preEditValidation, lineDelta, editStartLine } =
+        this.computeValidationContext(editData);
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/ast-edit/ast-edit-invocation.ts]: Fixed in db96b7c161: computeValidationContext now runs before displayResult construction, and structured metadata uses the same computed astValidation/preEditValidation values as the LLM summary. Added apply-path metadata assertions.
